File: frameworks/pytorch/multihead_elman.cc
// Copyright 2024 Erik Garrison. Apache 2.0 License.
//
// PyTorch bindings for Multi-head Elman RNN.

#include <ATen/cuda/CUDAContext.h>
#include <c10/cuda/CUDAGuard.h>
#include <torch/extension.h>
#include <vector>

#include "haste.h"
#include "support.h"

namespace {

using haste::v0::multihead_elman::ForwardPass;
using haste::v0::multihead_elman::BackwardPass;

using torch::Tensor;

std::vector<Tensor> multihead_elman_forward(
    bool training,
    Tensor x,           // [T, B, nheads, headdim] - input sequence
    Tensor h0,          // [B, nheads, headdim] - initial hidden state
    Tensor R,           // [nheads, headdim, headdim] - recurrent weights per head
    Tensor Wx,          // [nheads, headdim, headdim] - input weights per head
    Tensor bias,        // [nheads, headdim] - bias per head
    int activation) {   // 0=softsign, 1=tanh_residual, 2=tanh
    const auto time_steps = x.size(0);
    const auto batch_size = x.size(1);
    const auto nheads = x.size(2);
    const auto headdim = x.size(3);

    CHECK_INPUT(x);
    CHECK_INPUT(h0);
    CHECK_INPUT(R);
    CHECK_INPUT(Wx);
    CHECK_INPUT(bias);

    const auto options = x.options();
    const at::cuda::CUDAGuard guard(options.device_index());

    // Output tensors
    // h contains all hidden states: h[0] = h0, h[1..T] = outputs
    Tensor h = torch::empty({ time_steps + 1, batch_size, nheads, headdim }, options);
    Tensor y = torch::empty({ time_steps, batch_size, nheads, headdim }, options);
    Tensor pre_act = training
        ? torch::empty({ time_steps, batch_size, nheads, headdim }, options)
        : torch::empty({ 0 }, options);

    // Workspace tensors
    Tensor tmp_Rh = torch::empty({ batch_size, nheads, headdim }, options);
    Tensor tmp_Wxx = torch::empty({ time_steps, batch_size, nheads, headdim }, options);

    // Initialize h[0] with h0
    h[0] = h0;

    AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, x.scalar_type(), "multihead_elman_forward", ([&] {
        ForwardPass<typename native_type<scalar_t>::T> forward(
            training,
            batch_size,
            nheads,
            headdim,
            activation,
            at::cuda::getCurrentCUDABlasHandle(),
            at::cuda::getCurrentCUDAStream());

        forward.Run(
            time_steps,
            ptr<scalar_t>(R),
            ptr<scalar_t>(Wx),
            ptr<scalar_t>(bias),
            ptr<scalar_t>(x),
            ptr<scalar_t>(h),
            ptr<scalar_t>(y),
            training ? ptr<scalar_t>(pre_act) : nullptr,
            ptr<scalar_t>(tmp_Rh),
            ptr<scalar_t>(tmp_Wxx));
    }));

    return { h, y, pre_act };
}

std::vector<Tensor> multihead_elman_backward(
    Tensor x,           // [T, B, nheads, headdim] - input sequence
    Tensor R,           // [nheads, headdim, headdim] - recurrent weights
    Tensor Wx,          // [nheads, headdim, headdim] - input weights
    Tensor h,           // [T+1, B, nheads, headdim] - hidden states from forward
    Tensor pre_act,     // [T, B, nheads, headdim] - saved pre-activations
    Tensor dy,          // [T, B, nheads, headdim] - gradient of loss w.r.t. output
    int activation) {   // 0=softsign, 1=tanh_residual, 2=tanh
    const auto time_steps = x.size(0);
    const auto batch_size = x.size(1);
    const auto nheads = x.size(2);
    const auto headdim = x.size(3);

    CHECK_INPUT(x);
    CHECK_INPUT(R);
    CHECK_INPUT(Wx);
    CHECK_INPUT(h);
    CHECK_INPUT(pre_act);
    CHECK_INPUT(dy);

    const auto options = x.options();
    const at::cuda::CUDAGuard guard(options.device_index());

    // Output gradient tensors
    Tensor dx = torch::empty({ time_steps, batch_size, nheads, headdim }, options);
    Tensor dR = torch::zeros({ nheads, headdim, headdim }, options);
    Tensor dWx = torch::zeros({ nheads, headdim, headdim }, options);
    Tensor db = torch::zeros({ nheads, headdim }, options);
    Tensor dh0 = torch::zeros({ batch_size, nheads, headdim }, options);

    // Workspace
    Tensor tmp_dpre = torch::empty({ batch_size, nheads, headdim }, options);
    Tensor tmp_dh = torch::empty({ batch_size, nheads, headdim }, options);

    AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, x.scalar_type(), "multihead_elman_backward", ([&] {
        BackwardPass<typename native_type<scalar_t>::T> backward(
            batch_size,
            nheads,
            headdim,
            activation,
            at::cuda::getCurrentCUDABlasHandle(),
            at::cuda::getCurrentCUDAStream());

        backward.Run(
            time_steps,
            ptr<scalar_t>(R),
            ptr<scalar_t>(Wx),
            ptr<scalar_t>(x),
            ptr<scalar_t>(h),
            ptr<scalar_t>(pre_act),
            ptr<scalar_t>(dy),
            ptr<scalar_t>(dx),
            ptr<scalar_t>(dR),
            ptr<scalar_t>(dWx),
            ptr<scalar_t>(db),
            ptr<scalar_t>(dh0),
            ptr<scalar_t>(tmp_dpre),
            ptr<scalar_t>(tmp_dh));
    }));

    return { dx, dh0, dR, dWx, db };
}

}  // anonymous namespace

void multihead_elman_init(py::module& m) {
    m.def("multihead_elman_forward", &multihead_elman_forward, "MultiHeadElman forward",
          py::call_guard<py::gil_scoped_release>());
    m.def("multihead_elman_backward", &multihead_elman_backward, "MultiHeadElman backward",
          py::call_guard<py::gil_scoped_release>());
}

File: frameworks/pytorch/multihead_elman.py
# Copyright 2024 Erik Garrison. Apache 2.0 License.
"""Multi-head Elman RNN with per-head recurrence matrices.

Architecture per timestep per head:
    h_new[i] = activation(R[i] @ h[i] + Wx[i] @ x[i] + b[i])

Key insight: Each head has its own headdim×headdim R matrix,
giving 2048x more expressive recurrence than Mamba2's scalar decays.
"""

import torch
import torch.nn as nn
from torch.autograd import Function

import haste_pytorch_lib as lib


class MultiHeadElmanFunction(Function):
    @staticmethod
    def forward(ctx, training, x, h0, R, Wx, bias, activation):
        # x: [T, B, nheads, headdim]
        # h0: [B, nheads, headdim]
        # R: [nheads, headdim, headdim] - recurrent weights per head
        # Wx: [nheads, headdim, headdim] - input weights per head
        # bias: [nheads, headdim]
        # activation: 0=softsign, 1=tanh_residual, 2=tanh
        h, y, pre_act = lib.multihead_elman_forward(
            training,
            x.contiguous(),
            h0.contiguous(),
            R.contiguous(),
            Wx.contiguous(),
            bias.contiguous(),
            activation
        )
        if training:
            ctx.save_for_backward(x, R, Wx, h, pre_act)
            ctx.activation = activation

        # y is [T, B, nheads, headdim]
        return y, h[-1]  # Return output and final hidden state

    @staticmethod
    def backward(ctx, grad_output, grad_h_final):
        x, R, Wx, h, pre_act = ctx.saved_tensors
        activation = ctx.activation

        T, B, nheads, headdim = grad_output.shape

        # Combine gradient from output and final hidden state
        dy = grad_output.contiguous()
        # Add grad_h_final to dy[-1]
        if grad_h_final is not None:
            dy = dy.clone()
            dy[-1] = dy[-1] + grad_h_final

        dx, dh0, dR, dWx, db = lib.multihead_elman_backward(
            x, R, Wx, h, pre_act, dy, activation
        )

        return None, dx, dh0, dR, dWx, db, None


class MultiHeadElman(nn.Module):
    """
    Multi-head Elman RNN with per-head recurrence matrices.

    Unlike Mamba2 which uses scalar decays per head (total: 64 parameters),
    this uses full headdim×headdim R matrices per head, giving 4096x more
    expressive recurrence (262,144 recurrence parameters per layer).

    Architecture per timestep per head:
        h_new[i] = activation(R[i] @ h[i] + Wx[i] @ x[i] + b[i])

    Activation options:
        0 = softsign: x / (1 + |x|) - gradient-friendly, non-saturating
        1 = tanh_residual: x + tanh(x) - preserves gradients via skip
        2 = tanh: standard tanh
    """

    def __init__(self, input_size, hidden_size, nheads=64, activation=0, batch_first=False):
        super().__init__()
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.nheads = nheads
        self.headdim = hidden_size // nheads
        self.activation = activation
        self.batch_first = batch_first

        assert hidden_size % nheads == 0, \
            f"hidden_size ({hidden_size}) must be divisible by nheads ({nheads})"

        # Per-head weights
        # R: [nheads, headdim, headdim] - recurrent weights
        self.R = nn.Parameter(torch.empty(nheads, self.headdim, self.headdim))

        # Wx: [nheads, headdim, headdim] - input weights
        self.Wx = nn.Parameter(torch.empty(nheads, self.headdim, self.headdim))

        # bias: [nheads, headdim]
        self.bias = nn.Parameter(torch.zeros(nheads, self.headdim))

        self._init_weights()

    def _init_weights(self):
        # Xavier initialization for weights
        nn.init.xavier_uniform_(self.R)
        nn.init.xavier_uniform_(self.Wx)
        nn.init.zeros_(self.bias)

    def forward(self, x, h0=None):
        """
        Args:
            x: input tensor [T, B, D] or [B, T, D] if batch_first
            h0: initial hidden state [B, nheads, headdim], defaults to zeros

        Returns:
            output: [T, B, D] or [B, T, D] if batch_first
            h_final: [B, nheads, headdim] - final hidden state
        """
        if self.batch_first:
            x = x.transpose(0, 1)

        T, B, D = x.shape

        # Reshape input to [T, B, nheads, headdim]
        x_heads = x.view(T, B, self.nheads, self.headdim)

        if h0 is None:
            h0 = torch.zeros(B, self.nheads, self.headdim,
                           dtype=x.dtype, device=x.device)

        output, h_final = MultiHeadElmanFunction.apply(
            self.training, x_heads, h0,
            self.R, self.Wx, self.bias, self.activation
        )

        # Reshape output back to [T, B, D]
        output = output.view(T, B, D)

        if self.batch_first:
            output = output.transpose(0, 1)

        return output, h_final

    def extra_repr(self):
        return (f'input_size={self.input_size}, hidden_size={self.hidden_size}, '
                f'nheads={self.nheads}, headdim={self.headdim}, '
                f'activation={self.activation}')
